### DISCUSSION TITLE
DBのスケーラビリティ

### DISCUSSION THREAD

**Discussion by kakeru-one** [kakeru-one/web-engineering-roundtable#49] [Ideas]
DBのスケーラビリティ 
- アクセス過多
  - マスター/スレイブ構成
  - 垂直スケーリング
- データ量過多
  - 垂直スケーリング（ディスクの容量を増やす。）
  - パーティション
  - シャーディング

**Comment by kakeru-one** [comment]
## [水平シャーディング](https://railsguides.jp/active_record_multiple_databases.html#%E6%B0%B4%E5%B9%B3%E3%82%B7%E3%83%A3%E3%83%BC%E3%83%87%E3%82%A3%E3%83%B3%E3%82%B0)・垂直シャーディングとはなんですか？
    

    水平方向のシャーディングは行（レコード）を基準にテーブルを分割し、垂直方向のシャーディングは列（カラム）を基準にテーブルを分割する。
    
    水平方向のシャーディングとは、データベースを分割して各データベースサーバーの行数を減らしながら「シャード（shard）」全体で同じスキーマを維持すること。一般に「マルチテナント」シャーディングと呼ばれる。
    
    垂直方向のシャーディングは、クエリが単一の列を返すようなデータベースで有用である。
    
    ## メリット
    
    - 水平方向のスケーリングを強化することができる。
        - 作業負荷が分散され、システムの処理能力が向上し、より多くのリクエストに対応できるようになります。
    - クエリ応答時間の短縮ができる。
        - シャードは数行の行と列で構成されます。そのため、[[データベースのクエリ](https://kinsta.com/blog/query-monitor/)](https://kinsta.com/blog/query-monitor/)を処理する時間は短くなります。一方、シャード化されていないデータベースクエリでは、何百、何千もの行を検索する必要があります。
    - 障害時の信頼性向上
        - 各シャードが自律的に動作するため、ダウンが発生しても影響を受けるシャードは一部にとどまる。
    
    ## デメリット
    
    - 複数シャードにまたがるSQLのテーブル結合を行うと、ネットワークI/Oが発生するので処理に時間がかかり、パフォーマンスが低下する。
        - したがって、すべてのシャードにバランスよくデータが分散されるようなシャーディングの手法を選択する必要がある。
    
    ## デメリットをできる限り防ぐには？
    
    以下の手法を状況によって選択する。
    
    - キーベースのシャーディング
        - ハッシュ関数でシャーディングキーの値をハッシュ化し、その出力結果でデータを特定のシャードにマッピングし、分散していく方式。
            - これによってシャードごとの偏りはなくなる。
            - データを論理的にグループ化することはできないので、ある条件で一括取得するようなクエリを実行する場合は、複数のシャードに跨ってデータを読み込むことになってしまう。
    - レンジベースのシャーディング
        - 指定した値の範囲に応じてデータベースをシャーディングする。
        - レンジベースのシャーディングは、データベースのホットスポットを引き起こす可能性がある。
            - 例えば、シーケンシャルなIDで分散していた場合に、実際にインサートされないが、IDがインクリメントされることが多くなった場合は偏りが出る。
    - ディレクトリベースのシャーディング
        - 論理的に関連するデータを同じシャードにグループ化することができる。
            - 例えば、ユーザーを都道府県ごとにグループ化するなど。


## 参考文献
https://kinsta.com/jp/blog/database-sharding/

**Reply by kakeru-one** [reply]
## レプリケーションについて
https://zenn.dev/kimkiyong/scraps/55ac35a02f1257#:~:text=%E9%80%81%E4%BF%A1%E3%81%97%E5%A7%8B%E3%82%81%E3%82%8B-,%E3%83%87%E3%83%BC%E3%82%BF%E3%83%99%E3%83%BC%E3%82%B9%E3%83%AC%E3%83%97%E3%83%AA%E3%82%B1%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3,-%E3%82%AA%E3%83%AA%E3%82%B8%E3%83%8A%E3%83%AB(%E3%83%9E%E3%82%B9%E3%82%BF%E3%83%BC)/%E3%82%B3%E3%83%94%E3%83%BC

**Comment by kakeru-one** [comment]
実際にどうやってコピーするのか？

**Reply by kakeru-one** [reply]
レプリケーション
https://envader.plus/course/10/scenario/1096
https://www.system-exe.co.jp/dbexpert34/

**Comment by yuyakawai0411** [comment]
### 用語集
- スケールアウトする時の手段
  - パーティション
  - シャーディング
  - レプリケーション

### シャーディングの質問
シャーディングによって分割されたDB1とDB２があるとき、DB1、DB2どちらにアクセスするかは、どのレイヤーで管理するのか？ DB側で管理できるのか？ それともアプリケーション側で管理するのか？

### パーティションの質問
パーティションによって分割されたtable1とtable2があるとき、table1、table2どちらにアクセスするかは、どのレイヤーで管理するのか？ => table1とtable2を統合するViewを作ればいいのか？

### todo
- [ ] 実際にシャーディングとバーティションをやってみる
